### DISCUSSION TITLE
data model

### DISCUSSION THREAD

**Discussion by tfahringer** [alexzhornyak/SCXML-tutorial#31] [Q&A]
Is it correct that any <datamodel> in SCXML no matter at what state nesting level is seen everywhere in the entire <scxml> document as long it occurs in the dynamic extent of an element?

Question: is d1 in the following example also seen in state s2?

[dataModel1.txt](https://github.com/alexzhornyak/SCXML-tutorial/files/11355137/dataModel1.txt)

As the <datamodel> can be included in atomic states, compound states and <parallel> I wonder whether there is any notion of private  or  shared <datamodel>.   Can there be a private data model within a state that is only seen by this state and its siblings but not by anyone else?
Similarly can there be a shared data model that is only seen inside a <parallel> construct but not outside?

**Comment by alexzhornyak** [comment]
There are no private datamodels, once datamodel is declared it is seen everywhere whithin current state machine session, the only thing you can do is to declare it later using \<scxml\> element binding set to 'late'

https://github.com/alexzhornyak/SCXML-tutorial/blob/master/Doc/scxml.md#3-attribute-binding

**Reply by tfahringer** [reply]
but is d1 also seen in state s2 in this example?
[dataModel1.txt](https://github.com/alexzhornyak/SCXML-tutorial/files/11355455/dataModel1.txt)

**Reply by alexzhornyak** [reply]
In your example \<scxml\> attributes are hidden. By default binding is early, so d1 will be seen because will be initialized at the beginning

**Comment by tfahringer** [comment]
If I have a deeply nested compound state structure with some <state s1> deep inside such a hierarchy that contains a single 
<datamodel>
     <data ID = counter ...>
</datamodel>

Is ID then only seen everywhere in the overall <scxml> session as long as s1 is active? If s1 is not active then no one sees ID. Is this correct?

If this is correct then the only way to implement globally accessible data that is always accessible as long as the overall scxml session is active, if we place the datamodel right at the topmost scxml element. Is this correct?

**Reply by alexzhornyak** [reply]
Read please about binding attribute. By default it doesn't matter on which level of depth it is declared. It will be seen everywhere from the very statemachine beginning.

**Comment by tfahringer** [comment]
The data binding in 5.3.3 is not entirely clear on that. It provides information at what time data elements are created and given initial values. But there is no information as to how long they are seen.
My guess is that all data elements are then seen until the state machine stops to exist. Is this correct?

**Reply by alexzhornyak** [reply]
Yes, all data elements are seen till the state machine exists. If we are talking about script datamodels like Ecma, Lua etc. Data elements after declaration are becoming global variables of script interpreter, so will exist until interpreter will be destroyed
